src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,343 @@
+    bb += ((i >>> 8) & 0xff).toByte
+    bb += ((i >>> 16) & 0xff).toByte
+    bb += ((i >>> 24) & 0xff).toByte
+  }
+
+  def updateIntToBytesLE(bb: ArrayBuilder[Byte], i: Int, pos: Int) {
+    bb(pos) = (i & 0xff).toByte
+    bb(pos + 1) = ((i >>> 8) & 0xff).toByte
+    bb(pos + 2) = ((i >>> 16) & 0xff).toByte
+    bb(pos + 3) = ((i >>> 24) & 0xff).toByte
+  }
+
+  def stringToBytesWithShortLength(bb: ArrayBuilder[Byte], s: String): Int = {
+    val bytes = s.getBytes
+    val l = bytes.length
+    shortToBytesLE(bb, l)
+    bb ++= bytes
+    2 + l
+  }
+
+  def stringToBytesWithIntLength(bb: ArrayBuilder[Byte], s: String): Int = {
+    val bytes = s.getBytes
+    val l = bytes.length
+    intToBytesLE(bb, l)
+    bb ++= bytes
+    4 + l
+  }
+
+  private def resetIndex(index: Array[Int]) {
+    var i = 0
+    while (i < index.length) {
+      index(i) = i
+      i += 1
+    }
+  }
+
+  private def quickSortWithIndex(a: Array[Double], idx: Array[Int], start: Int, n: Int) {
+    def swap(i: Int, j: Int) {
+      val tmp = idx(i)
+      idx(i) = idx(j)
+      idx(j) = tmp
+    }
+
+    if (n <= 1)
+      return
+
+    val pivotIdx = start + n / 2
+    val pivot = a(idx(pivotIdx))
+    swap(pivotIdx, start + n - 1)
+
+    var left = start
+    var right = start + n - 1
+
+    while (left < right) {
+      if (a(idx(left)) >= pivot)
+        left += 1
+      else if (a(idx(right - 1)) < pivot)
+        right -= 1
+      else {
+        swap(left, right - 1)
+        left += 1
+        right -= 1
+      }
+    }
+
+    swap(left, start + n - 1)
+
+    quickSortWithIndex(a, idx, start, left - start)
+    quickSortWithIndex(a, idx, left + 1, n - (left - start + 1))
+  }
+
+  def sortedIndex(a: Array[Double], idx: Array[Int]) {
+    val n = a.length
+    assert(idx.length == n)
+    resetIndex(idx)
+    quickSortWithIndex(a, idx, 0, n)
+  }
+
+  def computeFractional(input: Array[Double], fractional: Array[Double]): Int = {
+    val n = input.length
+    var totalFractional = 0d
+    var i = 0
+    while (i < n) {
+      val x = input(i)
+      val f = x - x.floor
+      fractional(i) = f
+      totalFractional += f
+      i += 1
+    }
+
+    val F = (totalFractional + 0.5).toInt
+    assert(F >= 0 && F <= n)
+    F
+  }
+
+  def roundWithConstantSum(resized: Array[Double], fractional: Array[Double], index: Array[Int],
+    indexInverse: Array[Int], output: IntConsumer, F: Int, expectedSize: Long, keepLast: Boolean = true) {
+    val n = resized.length
+    assert(fractional.length == n && index.length == n && indexInverse.length == n)
+    sortedIndex(fractional, index)
+
+    var i = 0
+    while (i < n) {
+      indexInverse(index(i)) = i
+      i += 1
+    }
+
+    i = 0
+    var newSize = 0d
+    while (i < n) {
+      val r = if (indexInverse(i) < F) resized(i).ceil else resized(i).floor
+      if (keepLast || i != n - 1)
+        output += r.toUInt.intRep
+      newSize += r
+      i += 1
+    }
+    assert(newSize == expectedSize)
```

Comment:
I added a check to the caller of this function. It produces a warning message and treats the probabilities as missing if the `sum(GP) - 1 > 0.001`. I also added this check to the`exportGen` code. Not sure this is the best behavior. I would prefer the GenotypeView did the checking but that would require iterating through GP and computing sum when calling `hasGP`. But this would require accessing elements twice per genotype.